core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java

```diff
@@ -358,10 +358,20 @@ public long spill(long numBytes) throws IOException {
           return 0L;
         }
 
+        long released = 0L;
+        // longArray will not be used anymore, but may take significant amount of memory
+        // but never released until the task is complete.
+        if (longArray != null) {
+          released += longArray.memoryBlock().size();
+          freeArray(longArray);
+          longArray = null;
+        }
+        if (released >= numBytes) {
+          return released;
```

Comment:
IIUC we can only free this pointer array if we spill its related data pages. Is this short circuit safe to do?